### PR TITLE
Headless browser testing infrastructure

### PR DESCRIPTION
**Pull request by fitzgen**
This is very WIP, still a few issues:

* [x] webpack-dev-server spews tons of logs to console even when I try to pipe stdout/stderr
* [x] some of those logs seem to be errors, but I'm not sure why, and eventually it builds OK
* [x] the headless test doesn't pass yet. I think the bundle isn't working. At least, it isn't working when I manually reproduce in `target/generated-tests/*`. It seems like it is trying to require packages that are not in the bundle.

If anyone is interested in digging in, here is the output of a failed test run:

<details>

```
$ cargo test add_headless
    Blocking waiting for file lock on build directory
    Finished dev [unoptimized + debuginfo] target(s) in 3.49s
     Running target/debug/deps/all-63758bc3bf7e0984

running 1 test
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...
✖ ｢wdm｣: Hash: 17f457fcb1bd3011e62c
Version: webpack 4.11.1
Time: 246ms
Built at: 07/02/2018 6:12:53 PM
          Asset       Size  Chunks             Chunk Names
      bundle.js    273 KiB    main  [emitted]  main
    0.bundle.js   5.09 KiB       0  [emitted]
    2.bundle.js  594 bytes       2  [emitted]
  bundle.js.map    312 KiB    main  [emitted]  main
0.bundle.js.map   88 bytes       0  [emitted]
2.bundle.js.map   88 bytes       2  [emitted]
Entrypoint main = bundle.js bundle.js.map
[./out.js] 157 bytes {0} [built] [failed] [1 error]
[./run.js] 863 bytes {main} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 18 hidden modules

ERROR in ./out.js
Module build failed: Error: ENOENT: no such file or directory, open '/Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/out.js'
 @ ./run.js 2:23-38
 @ multi (webpack)-dev-server/client?http://localhost:8082 ./run.js

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8082 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
✖ ｢wdm｣: Hash: 8ef1ba17dbe76880f5e9
Version: webpack 4.11.1
Time: 273ms
Built at: 07/02/2018 6:12:53 PM
          Asset       Size  Chunks             Chunk Names
      bundle.js    273 KiB    main  [emitted]  main
    0.bundle.js   5.09 KiB       0  [emitted]
    2.bundle.js  594 bytes       2  [emitted]
  bundle.js.map    312 KiB    main✖ ｢wdm｣: Hash: 734d1b24850ba9d58d28
Version: webpack 4.11.1
Time: 257ms
Built at: 07/02/2018 6:12:53 PM
          Asset       Size  Chunks             Chunk Names
      bundle.js    273 KiB    main  [emitted]  main
    0.bundle.js   5.09 KiB       0  [emitted]
    2.bundle.js  594 bytes       2  [emitted]
  bundle.js.map    312 KiB    main  [emitted]  main
0.bundle.js.map   88 bytes       0  [emitted]
2.bundle.js.map   88 bytes       2  [emitted]
Entrypoint main = bundle.js bundle.js.map
[./out.js] 157 bytes {0} [built] [failed] [1 error]
[./run.js] 863 bytes {main} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 18 hidden modules

ERROR in ./out.js
Module build failed: Error: ENOENT: no such file or directory, open '/Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/out.js'
 @ ./run.js 2:23-38
 @ multi (webpack)-dev-server/client?http://localhost:8080 ./run.js

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8080 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
[39m  [emitted]  main
0.bundle.js.map   88 bytes       0  [emitted]
2.bundle.js.map   88 bytes       2  [emitted]
Entrypoint main = bundle.js bundle.js.map
[./out.js] 157 bytes {0} [built] [failed] [1 error]
[./run.js] 863 bytes {main} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 18 hidden modules

ERROR in ./out.js
Module build failed: Error: ENOENT: no such file or directory, open '/Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/out.js'
 @ ./run.js 2:23-38
 @ multi (webpack)-dev-server/client?http://localhost:8084 ./run.js

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8084 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
✖ ｢wdm｣: Hash: a46c3eb872ec151998f5
Version: webpack 4.11.1
Time: 304ms
Built at: 07/02/2018 6:12:53 PM
          Asset       Size  Chunks             Chunk Names
      bundle.js    273 KiB    main  [emitted]  main
    0.bundle.js   5.09 KiB       0  [emitted]
    2.bundle.js  594 bytes       2  [emitted]
  bundle.js.map    312 KiB    main  [emitted]  main
0.bundle.js.map   88 bytes       0  [emitted]
2.bundle.js.map   88 bytes       2  [emitted]
Entrypoint main = bundle.js bundle.js.map
[./out.js] 157 bytes {0} [built] [failed] [1 error]
[./run.js] 863 bytes {main} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 18 hidden modules

ERROR in ./out.js
Module build failed: Error: ENOENT: no such file or directory, open '/Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/out.js'
 @ ./run.js 2:23-38
 @ multi (webpack)-dev-server/client?http://localhost:8083 ./run.js

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8083 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
ℹ ｢wdm｣: Failed to compile.
✖ ｢wdm｣: Hash: b9aa713cd191265503dd
Version: webpack 4.11.1
Time: 336ms
Built at: 07/02/2018 6:12:53 PM
          Asset       Size  Chunks             Chunk Names
      bundle.js    273 KiB    main  [emitted]  main
    0.bundle.js   5.09 KiB       0  [emitted]
    2.bundle.js  594 bytes       2  [emitted]
  bundle.js.map    312 KiB    main  [emitted]  main
0.bundle.js.map   88 bytes       0  [emitted]
2.bundle.js.map   88 bytes       2  [emitted]
Entrypoint main = bundle.js bundle.js.map
[./out.js] 157 bytes {0} [built] [failed] [1 error]
[./run.js] 863 bytes {main} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 18 hidden modules

ERROR in ./out.js
Module build failed: Error: ENOENT: no such file or directory, open '/Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/out.js'
 @ ./run.js 2:23-38
 @ multi (webpack)-dev-server/client?http://localhost:8081 ./run.js

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8081 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...
ℹ ｢wdm｣: Compiling...

> @ run-webpack-dev-server /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0
> webpack-dev-server

✖ ｢wdm｣: Hash: 77a565f9f0100e0d2cb9
Version: webpack 4.11.1
Time: 427ms
Built at: 07/02/2018 6:12:56 PM
          Asset      Size  Chunks             Chunk Names
      bundle.js   276 KiB    main  [emitted]  main
    0.bundle.js  67.7 KiB       0  [emitted]
  bundle.js.map   315 KiB    main  [emitted]  main
0.bundle.js.map  43.5 KiB       0  [emitted]
 + 3 hidden assets
Entrypoint main = bundle.js bundle.js.map
[./out.js] 37.4 KiB {0} [built]
[./out_bg.wasm] 233 KiB {0} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 20 hidden modules

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8084 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
✖ ｢wdm｣: Hash: 041453af0072571819f7
Version: webpack 4.11.1
Time: 434ms
Built at: 07/02/2018 6:12:56 PM
          Asset      Size  Chunks             Chunk Names
      bundle.js   276 KiB    main  [emitted]  main
    0.bundle.js  67.7 KiB       0  [emitted]
  bundle.js.map   315 KiB    main  [emitted]  main
0.bundle.js.map  43.5 KiB       0  [emitted]
 + 3 hidden assets
Entrypoint main = bundle.js bundle.js.map
[./out.js] 37.4 KiB {0} [built]
[./out_bg.wasm] 233 KiB {0} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 20 hidden modules

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8080 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
✖ ｢wdm｣: Hash: 24ede76bc0586071a1f7
Version: webpack 4.11.1
Time: 445ms
Built at: 07/02/2018 6:12:56 PM
          Asset      Size  Chunks             Chunk Names
      bundle.js   276 KiB    main  [emitted]  main
    0.bundle.js  67.7 KiB       0  [emitted]
  bundle.js.map   315 KiB    main  [emitted]  main
0.bundle.js.map  43.5 KiB       0  [emitted]
 + 3 hidden assets
Entrypoint main = bundle.js bundle.js.map
[./out.js] 37.4 KiB {0} [built]
[./out_bg.wasm] 233 KiB {0} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 20 hidden modules

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8081 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
✖ ｢wdm｣: Hash: 19682aa16b77d78508eb
Version: webpack 4.11.1
Time: 449ms
Built at: 07/02/2018 6:12:56 PM
          Asset      Size  Chunks             Chunk Names
      bundle.js   276 KiB    main  [emitted]  main
    0.bundle.js  67.7 KiB       0  [emitted]
  bundle.js.map   315 KiB    main  [emitted]  main
0.bundle.js.map  43.5 KiB       0  [emitted]
 + 3 hidden assets
Entrypoint main = bundle.js bundle.js.map
[./out.js] 37.4 KiB {0} [built]
[./out_bg.wasm] 233 KiB {0} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 20 hidden modules

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8082 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
✖ ｢wdm｣: Hash: 8015b31c3893ec3ba596
Version: webpack 4.11.1
Time: 455ms
Built at: 07/02/2018 6:12:56 PM
          Asset      Size  Chunks             Chunk Names
      bundle.js   276 KiB    main  [emitted]  main
    0.bundle.js  67.7 KiB       0  [emitted]
  bundle.js.map   315 KiB    main  [emitted]  main
0.bundle.js.map  43.5 KiB       0  [emitted]
 + 3 hidden assets
Entrypoint main = bundle.js bundle.js.map
[./out.js] 37.4 KiB {0} [built]
[./out_bg.wasm] 233 KiB {0} [built]
[./test.ts] 348 bytes {2} [built] [failed] [2 errors]
    + 20 hidden modules

ERROR in ./test.ts
Module build failed: Error: Typescript emitted no output for /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts.
    at successLoader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:41:15)
    at Object.loader (/Users/fitzgen/src/wasm-bindgen/node_modules/ts-loader/dist/index.js:21:12)
 @ ./run.js 3:25-41
 @ multi (webpack)-dev-server/client?http://localhost:8083 ./run.js

ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts
./test.ts
[tsl] ERROR in /Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/test.ts(3,35)
      TS2307: Cannot find module './out'.
ℹ ｢wdm｣: Failed to compile.
ℹ ｢wdm｣: Failed to compile.
ℹ ｢wds｣: Project is running at http://localhost:8085/
ℹ ｢wds｣: webpack output is served from /
ℹ ｢wdm｣: Hash: 5685e6eace2969b12da7
Version: webpack 4.11.1
Time: 1598ms
Built at: 07/02/2018 6:12:58 PM
                           Asset          Size  Chunks             Chunk Names
                       bundle.js       276 KiB    main  [emitted]  main
                     0.bundle.js      67.7 KiB       0  [emitted]
1ff1c8c7e96d317ef4f1.module.wasm  unknown size       0  [emitted]
                     1.bundle.js     608 bytes       1  [emitted]
                   bundle.js.map       316 KiB    main  [emitted]  main
                 0.bundle.js.map      43.5 KiB       0  [emitted]
                 1.bundle.js.map     513 bytes       1  [emitted]
Entrypoint main = bundle.js bundle.js.map
[../../../node_modules/ansi-html/index.js] /Users/fitzgen/src/wasm-bindgen/node_modules/ansi-html/index.js 4.16 KiB {main} [built]
[../../../node_modules/html-entities/index.js] /Users/fitzgen/src/wasm-bindgen/node_modules/html-entities/index.js 231 bytes {main} [built]
[../../../node_modules/loglevel/lib/loglevel.js] /Users/fitzgen/src/wasm-bindgen/node_modules/loglevel/lib/loglevel.js 7.68 KiB {main} [built]
[../../../node_modules/webpack-dev-server/client/index.js?http://localhost:8085] (webpack)-dev-server/client?http://localhost:8085 7.75 KiB {main} [built]
[../../../node_modules/webpack-dev-server/client/overlay.js] (webpack)-dev-server/client/overlay.js 3.58 KiB {main} [built]
[../../../node_modules/webpack-dev-server/client/socket.js] (webpack)-dev-server/client/socket.js 1.05 KiB {main} [built]
[../../../node_modules/webpack-dev-server/node_modules/strip-ansi/index.js] (webpack)-dev-server/node_modules/strip-ansi/index.js 161 bytes {main} [built]
[../../../node_modules/webpack/hot sync ^\.\/log$] (webpack)/hot sync nonrecursive ^\.\/log$ 170 bytes {main} [built]
[../../../node_modules/webpack/hot/emitter.js] (webpack)/hot/emitter.js 77 bytes {main} [built]
[../../../node_modules/webpack/hot/log.js] (webpack)/hot/log.js 1010 bytes {main} [optional] [built]
[./out.js] 37.4 KiB {0} [built]
[./run.js] 863 bytes {main} [built]
[./test.ts] 286 bytes {1} [built]
[url] external "url" 42 bytes {main} [built]
[0] multi (webpack)-dev-server/client?http://localhost:8085 ./run.js 40 bytes {main} [built]
    + 9 hidden modules
ℹ ｢wdm｣: Compiled successfully.
test simple::add_headless ... FAILED

failures:

---- simple::add_headless stdout ----
···················································
running "cargo" "build" "--target" "wasm32-unknown-unknown"
exit: exit code: 0
dur: 0.847ms
stderr ---
   Compiling test0 v0.0.1 (file:///Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0)
    Finished dev [unoptimized + debuginfo] target(s) in 0.81s

···················································
running "node" "/Users/fitzgen/src/wasm-bindgen/target/generated-tests/test0/run-headless.js"
exit: exit code: 1
dur: 8.127ms
stderr ---
Got an error: TimeoutError: Waiting until element text is
Wait timed out after 6002ms

Stack: TimeoutError: Waiting until element text is
Wait timed out after 6002ms
    at /Users/fitzgen/src/wasm-bindgen/node_modules/selenium-webdriver/lib/webdriver.js:834:17
    at process._tickCallback (internal/process/next_tick.js:68:7)

thread 'simple::add_headless' panicked at 'assertion failed: output.status.success()', tests/all/main.rs:692:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
···················································
background webpack-dev-server
stdout ---

stderr ---



failures:
    simple::add_headless

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 227 filtered out

error: test failed, to rerun pass '--test all'
```

</details>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,14 +1,19 @@
 {
   "license": "MIT",
   "scripts": {
-    "run-webpack": "webpack"
+    "run-webpack": "webpack",
+    "run-webpack-dev-server": "webpack-dev-server",
```
Comment (sendilkumarn) [inline package.json]: can we rename it to `serve` ? That is what everybody uses

File: tests/all/main.rs [hunk 2]
```diff
@@ -264,28 +277,30 @@ impl Project {
     }
 
     fn generate_js_entry(&mut self, modules: Vec<PathBuf>) {
-        let mut runjs = r#"
-            import * as process from "process";
-        "#.to_string();
+        let mut runjs = if self.headless {
+            "".to_string()
+        } else {
+            "import * as process from \"process\";".to_string()
```
Comment (sendilkumarn) [inline tests/all/main.rs]: `r#` was more expressive for me 😄 

Is there any advantage using this over raw string ?

**Comment by alexcrichton** [comment]
I've pushed a few commits that should fix running webpack-dev-server in the background as well as get the test passing and get test failures a bit more legible (aka spewing logs and errors and such to the test logs)

File: tests/all/main.rs (hunk 2 above)
Comment (fitzgen) [reply tests/all/main.rs]: My highlighting was starting to freak out and I just did that as a temporary fix. Agreed that `r#` strings are better here.

File: tests/all/main.rs [hunk 3]
```diff
@@ -300,21 +323,42 @@ impl Project {
...
-                if (wasm.assertSlabEmpty)
-                    wasm.assertSlabEmpty();
-            })
-            .catch(error => {
-                  console.error(error);
-                  process.exit(1);
-                });
-            "#,
+                .then(results => {
+                    let [test, wasm] = results;
```
Comment (sendilkumarn) [inline tests/all/main.rs]: can be `const` here

File: tests/all/main.rs [hunk 4]
```diff
@@ -300,21 +323,42 @@ impl Project {
...
+            "#
         );
+
+        if self.headless {
+            runjs.push_str(
+                r#"
+                    .then(() => document.getElementById('status').innerHTML = 'good')
+                    .catch(e => {
+                        const errors = document.getElementById('error');
+                        let content = `exception: ${e.message}\nstack: ${e.stack}`;
```
Comment (sendilkumarn) [inline tests/all/main.rs]: `const` here too

File: tests/all/main.rs [hunk 5]
```diff
@@ -300,21 +323,42 @@ impl Project {
...
+                        errors.innerHTML = `<pre>${content}</pre>`;
+                    })
+                    .finally(() => {
+                        window.document.body.innerHTML += "\n TESTDONE";
```
Comment (sendilkumarn) [inline tests/all/main.rs]: `document` will be enough right?

File: tests/all/main.rs [hunk 6]
```diff
@@ -457,6 +506,146 @@ impl Project {
...
+                async function main() {
+                    const body = driver.findElement(By.tagName('body'));
+                    try {
+                        await driver.get('http://localhost:8080/index.html');
+                        await driver.wait(
+                            until.elementTextContains(body, 'TESTDONE'),
+                            6 * 1000
+                        );
+
+                        let status = await body.findElement(By.id('status')).getText();
```
Comment (sendilkumarn) [inline tests/all/main.rs]: `status` | `log` | `errors` can be `const`

**Comment by fitzgen** [comment]
Crap, CI started failing again :(

```
failures:
---- simple::add_headless stdout ----
···················································
running "cargo" "build" "--target" "wasm32-unknown-unknown"
exit: exit code: 0
dur: 59.309ms
stderr ---
    Blocking waiting for file lock on build directory
   Compiling test219 v0.0.1 (file:///home/travis/build/rustwasm/wasm-bindgen/target/generated-tests/test219)
    Finished dev [unoptimized + debuginfo] target(s) in 58.81s
···················································
running "node" "/home/travis/build/rustwasm/wasm-bindgen/target/generated-tests/test219/run-headless.js"
exit: exit code: 1
dur: 26.470ms
stderr ---
Got an error: TimeoutError: Waiting until element text contains
Wait timed out after 6003ms
Stack: TimeoutError: Waiting until element text contains
Wait timed out after 6003ms
    at /home/travis/build/rustwasm/wasm-bindgen/node_modules/selenium-webdriver/lib/webdriver.js:834:17
    at process._tickCallback (internal/process/next_tick.js:68:7)
thread 'simple::add_headless' panicked at 'assertion failed: output.status.success()', tests/all/main.rs:709:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:221
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:511
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:445
   6: all::run
             at tests/all/main.rs:709
   7: all::Project::test_headless
             at tests/all/main.rs:565
   8: all::Project::test
             at tests/all/main.rs:518
   9: all::simple::add_headless
             at tests/all/simple.rs:64
  10: all::__test::TESTS::{{closure}}
             at tests/all/simple.rs:63
  11: core::ops::function::FnOnce::call_once
             at /checkout/src/libcore/ops/function.rs:223
  12: <F as alloc::boxed::FnBox<A>>::call_box
             at libtest/lib.rs:1451
             at /checkout/src/libcore/ops/function.rs:223
             at /checkout/src/liballoc/boxed.rs:640
  13: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:105
···················································
background webpack-dev-server
status: exit code: 0
stdout ---
> @ run-webpack-dev-server /home/travis/build/rustwasm/wasm-bindgen/target/generated-tests/test219
> webpack-dev-server "--quiet" "--watch-stdin"
stderr ---
failures:
    simple::add_headless
```

**Comment by alexcrichton** [comment]
This looks good to me, feel free to merge when CI is green!

**Comment by alexcrichton** [comment]
The latest commit [successfully executed](https://travis-ci.com/alexcrichton/wasm-bindgen) on my local Travis, no idea how that last commit could have fixed anything. Let's see how it fares with this repo's travis...

**Comment by alexcrichton** [comment]
Magically this is now green? No idea! Seems reproducible though, so let's merge!

**Comment by fitzgen** [comment]
Yay? Yay!

**Pull request merged**